File: core/src/main/java/javabot/dao/util/LocalDateTimeConverter.java
package javabot.dao.util;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.Date;

import org.mongodb.morphia.converters.DateConverter;
import org.mongodb.morphia.mapping.MappedField;
import org.mongodb.morphia.mapping.MappingException;

public class LocalDateTimeConverter extends DateConverter {
  public LocalDateTimeConverter() {
    super(LocalDateTime.class);
  }

  @Override
  public Object decode(final Class targetClass, final Object val, final MappedField optionalExtraInfo)
      throws MappingException {
    Date date = (Date) super.decode(targetClass, val, optionalExtraInfo);
    Instant instant = Instant.ofEpochMilli(date.getTime());
    return LocalDateTime.ofInstant(instant, ZoneOffset.UTC);
  }

  @Override
  public Object encode(final Object value, final MappedField optionalExtraInfo) {
    if (value == null) {
      return null;
    }
    return new Date(((LocalDateTime) value).toInstant(ZoneOffset.UTC).toEpochMilli());
  }
}

File: src/main/kotlin/javabot/dao/util/LocalDateTimeConverter.kt
package javabot.dao.util

import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneOffset
import java.util.Date

import dev.morphia.converters.DateConverter
import dev.morphia.mapping.MappedField
import dev.morphia.mapping.MappingException

class LocalDateTimeConverter : DateConverter(LocalDateTime::class.java) {

    @Throws(MappingException::class)
    override fun decode(targetClass: Class<*>?, value: Any?, optionalExtraInfo: MappedField?): Any? {
        val date = super.decode(targetClass, value, optionalExtraInfo) as Date
        val instant = Instant.ofEpochMilli(date.time)
        return LocalDateTime.ofInstant(instant, ZoneOffset.UTC)
    }

    override fun encode(value: Any?, optionalExtraInfo: MappedField?): Any? {
        if (value == null) {
            return null
        }
        return Date((value as LocalDateTime).toInstant(ZoneOffset.UTC).toEpochMilli())
    }
}
